File: ice0208/two-pointers/11.py
class Solution:
    def maxArea(self, height: list[int]) -> int:
        l, r = 0, len(height) - 1

        answer = 0
        while l < r:
            w = r - l
            h = min(height[l], height[r])

            answer = max(answer, w * h)

            if height[l] < height[r]:
                l += 1
            else:
                r -= 1

        return answer


if __name__ == "__main__":
    print(Solution().maxArea(height=[1, 8, 6, 2, 5, 4, 8, 3, 7]))  # 49
    print(Solution().maxArea(height=[1, 1]))  # 1

File: moonpoison/two_pointers/11.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int maxArea(vector<int>& height) 
    {
        int ans=0;
        int x1=0, x2=height.size()-1;
        while(x1<x2)
        {
            int x=x2-x1;
            int y=min(height[x1], height[x2]);
            int area=x*y;
            if(ans<area) ans=area;
            if(height[x1]<height[x2]) x1++;
            else x2--;
        }
        return ans;
    }
};
